From d28ef6c5ae480d6e0f6f50d341d534e01cc6c307 Mon Sep 17 00:00:00 2001
From: Robert Lai <c_rlai@codeaurora.org>
Date: Mon, 31 Aug 2015 18:06:06 -0700
Subject: [PATCH] audio: Fix stale session opening of DSP audio decoder

- Device crash observed on stability test due to
  max asm session open failure

- While reject DSP audio NT decoders, sessions are
  opened before checking max limit & error is posted
  to frameworks. This results in stale session.

- On MAX session limit, reject DSP NT decoders
  without creating component

Change-Id: I8c28c264a2953b1f80051658e053dd121249041c
---
 mm-core/src/common/qc_omx_core.c | 22 ++++++++++++++--------
 1 file changed, 14 insertions(+), 8 deletions(-)

diff --git a/mm-core/src/common/qc_omx_core.c b/mm-core/src/common/qc_omx_core.c
index 3564396..57f8148 100644
--- a/mm-core/src/common/qc_omx_core.c
+++ b/mm-core/src/common/qc_omx_core.c
@@ -427,6 +427,17 @@ OMX_GetHandle(OMX_OUT OMX_HANDLETYPE*     handle,
 
       if(core[cmp_index].fn_ptr)
       {
+        //Do not allow more than MAX limit for DSP audio decoders
+        if((!strcmp(core[cmp_index].so_lib_name,"libOmxWmaDec.so")  ||
+            !strcmp(core[cmp_index].so_lib_name,"libOmxAacDec.so")  ||
+            !strcmp(core[cmp_index].so_lib_name,"libOmxAlacDec.so") ||
+            !strcmp(core[cmp_index].so_lib_name,"libOmxApeDec.so")) &&
+            (number_of_adec_nt_session+1 > MAX_AUDIO_NT_SESSION)) {
+            DEBUG_PRINT_ERROR("Rejecting new session..Reached max limit for DSP audio decoder session");
+            core[cmp_index].inst[hnd_index]= *handle = NULL;
+            pthread_mutex_unlock(&lock_core);
+            return OMX_ErrorInsufficientResources;
+        }
         // Construct the component requested
         // Function returns the opaque handle
         void* pThis = (*(core[cmp_index].fn_ptr))();
@@ -459,14 +470,9 @@ OMX_GetHandle(OMX_OUT OMX_HANDLETYPE*     handle,
              !strcmp(core[cmp_index].so_lib_name,"libOmxAacDec.so")  ||
              !strcmp(core[cmp_index].so_lib_name,"libOmxAlacDec.so") ||
              !strcmp(core[cmp_index].so_lib_name,"libOmxApeDec.so")) {
-            if(number_of_adec_nt_session+1 > MAX_AUDIO_NT_SESSION) {
-              DEBUG_PRINT_ERROR("Audio NT session max limit is 2\n");
-              core[cmp_index].inst[hnd_index]= *handle = NULL;
-              pthread_mutex_unlock(&lock_core);
-              return OMX_ErrorInsufficientResources;
-            }
-          number_of_adec_nt_session++;
-          DEBUG_PRINT_ERROR("OMX_GetHandle: number_of_adec_nt_session : %d\n",
+
+             number_of_adec_nt_session++;
+             DEBUG_PRINT("OMX_GetHandle: number_of_adec_nt_session : %d\n",
                              number_of_adec_nt_session);
           }
         }
-- 
1.9.1

